Assets/Scripts/Cheat/Main/TowerController.cs

```diff
@@ -0,0 +1,125 @@
+﻿/*
+ * Author(s): Isaiah Mann
+ * Description: Controls all the towers in the game
+ */
+
+using UnityEngine;
+using SimpleJSON;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine.EventSystems;
+
+public class TowerController : UnitController<ITower, Tower, TowerList>, ITowerController {
+	public const string TOWER_TAG = "Tower";
+	const float DRAG_HEIGHT_OFFSET = 1.25f;
+
+	public GameObject CoreOrbPrefab;
+	public GameObject CoreOrbInstance;
+	GameObject potentialPurchaseTower = null;
+	MapTileBehaviour previousHighlightedMapTile = null;
+	HashSet<TowerBehaviour> activeTowers = new HashSet<TowerBehaviour>();
+	public static List<TowerBehaviour> healTowers = new List<TowerBehaviour> (); //9/29/16
+
+	public override void Setup (WorldController worldController, DataController dataController, string unitTemplateJSONPath) {
+		base.Setup(worldController, dataController, unitTemplateJSONPath);
+	}
+		
+	public void HandleBeginDragPurchase (PointerEventData dragEvent, TowerPurchasePanel towerPanel) {
+		if (potentialPurchaseTower != null) {
+			Destroy(potentialPurchaseTower);
+		}
+		potentialPurchaseTower = (GameObject)Instantiate(worldController.GetTowerPrefab(towerPanel.TowerType));
+		potentialPurchaseTower.GetComponent<TowerBehaviour>().ToggleColliders(false);
+	}
+
+	public void HandleDragPurchase (PointerEventData dragEvent, TowerPurchasePanel towerPanel) {
+		Vector3 dragPosition = dragEvent.position;
+		dragPosition.z = towerPanel.transform.position.z - Camera.main.transform.position.z;
+		dragPosition = Camera.main.ScreenToWorldPoint(dragPosition);
+		dragPosition.y = DRAG_HEIGHT_OFFSET;
+		potentialPurchaseTower.transform.position = dragPosition;
+		HighlightSpotToPlace(dragPosition);
+	}
+
+	void HighlightSpotToPlace (Vector3 dragPosition) {
+		RaycastHit hit;
+		if (Physics.Raycast(dragPosition, Vector3.down, out hit)) {
+			MapTileBehaviour mapTile;
+			if (hit.collider != null && (mapTile = hit.collider.GetComponent<MapTileBehaviour>()) != null) {
+				if (previousHighlightedMapTile) {
+					previousHighlightedMapTile.Unhighlight();
+				}
+				if (!mapTile.HasAgent()) {
+					mapTile.HightlightToPlace(potentialPurchaseTower.GetComponent<StaticAgentBehaviour>());
+					previousHighlightedMapTile = mapTile;
+				}
+			}
+		}
+	}
+
+	public void AddActiveTower (TowerBehaviour tower) {
+		if (!activeTowers.Contains(tower)) {
+			activeTowers.Add(tower);
+			healTowers.Add (tower);	//9/29/16
+		}
+	}
+
+	public void RemoveActiveTower (TowerBehaviour tower) {
+		if (activeTowers.Contains(tower)) {
+			activeTowers.Remove(tower);
+			healTowers.Add (tower);	//9/29/16
```

Comment:
See previous note about using the HashSet instead of creating a new list, but also unclear why we're adding the tower to the list again (this is typically when the tower is destroyed)